api/insert/insert.go

```diff
@@ -0,0 +1,180 @@
+	if err != nil {
+		return err
+	}
+	r.Body.Close()
+
+	var req jsonAddRequest
+
+	err = json.Unmarshal(body, &req)
+	if err != nil {
+		return errors.NewBadRequestString("Unable to parse certificate addition request")
+	}
+
+	if len(req.Serial) == 0 {
+		return errors.NewBadRequestString("Serial number is required but not provided")
+	}
+
+	if len(req.AKI) == 0 {
+		return errors.NewBadRequestString("Authority key identifier is required but not provided")
+	}
+
+	if _, present := ocsp.StatusCode[req.Status]; !present {
+		return errors.NewBadRequestString("Invalid certificate status")
+	}
+
+	if ocsp.StatusCode[req.Status] == stdocsp.Revoked && req.RevokedAt == (time.Time{}) {
+		return errors.NewBadRequestString("Revoked certificate should specify when it was revoked")
+	}
+
+	if _, present := validReasons[req.Reason]; !present {
+		return errors.NewBadRequestString("Invalid certificate status reason code")
+	}
+
+	if len(req.PEM) == 0 {
+		return errors.NewBadRequestString("The provided certificate is empty")
+	}
+
+	// Parse the certificate and validate that it matches
+	cert, err := helpers.ParseCertificatePEM([]byte(req.PEM))
+	if err != nil {
+		return errors.NewBadRequestString("Unable to parse PEM encoded certificates")
+	}
+
+	serialBigInt := new(big.Int)
+	if _, success := serialBigInt.SetString(req.Serial, 16); !success {
+		return errors.NewBadRequestString("Unable to parse serial key of request")
+	}
+
+	if serialBigInt.Cmp(cert.SerialNumber) != 0 {
+		return errors.NewBadRequestString("Serial key of request and certificate do not match")
+	}
+
+	aki, err := hex.DecodeString(req.AKI)
+	if err != nil {
+		return errors.NewBadRequestString("Unable to decode authority key identifier")
+	}
+
+	if !bytes.Equal(aki, cert.AuthorityKeyId) {
+		return errors.NewBadRequestString("Authority key identifier of request and certificate do not match")
+	}
+
+	cr := certdb.CertificateRecord{
+		Serial:    req.Serial,
+		AKI:       req.AKI,
+		CALabel:   req.CALabel,
+		Status:    req.Status,
+		Reason:    req.Reason,
+		Expiry:    req.Expiry,
+		RevokedAt: req.RevokedAt,
+		PEM:       req.PEM,
+	}
+
+	err = h.dbAccessor.InsertCertificate(cr)
+	if err != nil {
+		return err
+	}
+
+	if h.signer != nil {
+		// Now create an appropriate OCSP response
+		sr := ocsp.SignRequest{
+			Certificate: cert,
+			Status:      req.Status,
+			Reason:      req.Reason,
+			RevokedAt:   req.RevokedAt,
+		}
+		ocspResponse, err := h.signer.Sign(sr)
+		if err != nil {
+			return err
+		}
+
+		// We parse the OCSP repsonse in order to get the next
+		// update time/expiry time
+		ocspParsed, err := stdocsp.ParseResponse(ocspResponse, nil)
+		if err != nil {
+			return err
+		}
+
+		ocspRecord := certdb.OCSPRecord{
+			Serial: req.Serial,
+			AKI:    req.AKI,
+			Body:   string(ocspResponse),
+			Expiry: ocspParsed.NextUpdate,
+		}
+
+		if err = h.dbAccessor.InsertOCSP(ocspRecord); err != nil {
+			return err
+		}
+	}
+
+	result := map[string]string{}
```

Comment:
we may want to return {"certificate": the pem string, "ocsp_response": base64_encoded(ocspResponse)}, it gives API user a good UX, also it can enrich our testing by inspecting the return json object.